toolkit.nu

```diff
@@ -211,3 +211,41 @@ export def "check pr" [
 
     report --no-fail
 }
+
+# creates git hooks to run:
+# - `toolkit fmt` on `git commit`
+# - `toolkit fmt --check` and `toolkit clippy` on `git push`
+export def set-git-hooks [] {
+    set-git-hook pre-commit `#!/usr/bin/env nu
+
+use ../../toolkit.nu [fmt, pretty-print-command]
+
+print $"running ('toolkit fmt' | pretty-print-command)"
+fmt`
+
+    set-git-hook pre-push `#!/usr/bin/env nu
+
+use ../../toolkit.nu [fmt, clippy, pretty-print-command]
+
+print $"running ('toolkit fmt' | pretty-print-command)"
+try {
+    fmt --check
+} catch {
+    print $"\nplease run (ansi default_dimmed)(ansi default_italic)toolkit fmt(ansi reset) to fix the formatting"
+    exit 1
+}
```

Comment:
no need for this `try / catch` => the `cargo fmt --all -- --check` in `toolkit fmt` already returns this exit code :wink: 

e.g. if you add bad formatting
```
toolkit fmt --check
print 'foo'
```
would never print 'foo'

Reply:
@amtoine Right, the reason I added this is to add that last message to the output error (`"please run toolkit fmt to fix the formatting"`) so that the user knows what the above output is and how to fix it. Else only the formatting diff is displayed.

Reply:
:scream: you are right, it's not part of `toolkit fmt --check`

i have that line in another hook in another project :yum: 

i'm wondering if that should be part of the `fmt` and `clippy` command instead of being "scattered" around :thinking: 

that's a bit out of the scope of that PR of yours, so feel free to
- leave it as-is, in `toolkit setup`
- move these in `fmt` and `clippy`

i'm fine with either of them, even though that might lead to a little new `--verbose` option on these two commands :smirk:

Reply:
I'm happy to include the `--verbose` functionality as part of this PR. I'm already changing the toolkit's public API so might as well include that too.

Reply:
looking forward to it then :relieved:

Reply:
Silly question, but what would be the right way to return an error inside a catch block? I'm unsure of how to move the try-catch to `fmt` and `clippy` but still print the reports in `"check pr"`... 😕

Reply:
As in I want to add the extra error messages, but I also want the commands to fail when the corresponding `cargo` commands fail.

Reply:
`error make` might do the trick, right?
that should print a message AND quit at the same time.

you can check the `help error make` examples to see the possible format :wink:

Reply:
Yep I considered it but then it looks kinda clunky since there's not really a relevant span to use, so it's just a single message. It works though.

Reply:
`error make --unspanned` is your thing then :+1:

Reply:
i think it looks great
```bash
>_ toolkit fmt --check
Diff in /home/amtoine/.local/share/git/store/github.com/amtoine/nushell/crates/nu-command/src/platform/ansi/ansi_.rs at line 12:
 pub struct AnsiCommand;

 struct AnsiCode {
-
     short_name: Option<&'static str>,
     long_name: &'static str,
     code: String,
Error:
  × please run `toolkit fmt`
```

Reply:
In that case, are we okay with always having an error message despite `--verbose` being false? Then I don't see much value in the `--verbose` parameter...

Reply:
i can't think of a use case right now where we would want the `toolkit` to only be `--verbose` without `error make`ing :thinking: 

so it'd say let's go with `error make` and no `--verbose` :muscle: 
that can always be changed later.